<img src="https://raw.githubusercontent.com/alan-barzilay/NLPortugues/master/imagens/logo_nlportugues.png"   width="150" align="right">

# Lista 4 - Word2Vec 

______________

Nessa lista nós exploraremos o espaço vetorial gerado pelo algoritmo Word2Vec e algumas de suas propriedades mais interessantes. Veremos como palavras similares se organizam nesse espaço e as relações de palavras com seus sinônimos e antônimos. Também veremos algumas analogias interessantes que o algoritmo é capaz de fazer ao capturar um pouco do nosso uso da língua portuguesa.


In [1]:
from gensim.models import KeyedVectors

# Carregando dados


Para esta lista nós utilizaremos vetores de palavras, também conhecidos como *embeddings*, para lingua portuguesa fornecidos pelo [NILC](http://www.nilc.icmc.usp.br/nilc/index.php). Nós utilizaremos o embedding de 50 dimensões treinado com o algoritmo Word2Vec (Continous Bag of Words) que pode ser encontrado [aqui](http://www.nilc.icmc.usp.br/embeddings) sob a licensa [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/). Para evitar problemas de mémoria utilizaremos apenas as 200 mil palavras mais comum.

In [ ]:
!curl  https://raw.githubusercontent.com/alan-barzilay/NLPortugues/master/Semana%2004/data/word2vec_200k.txt --output 'word2vec_200k.txt'

In [2]:
# Carrega word2vec
model = KeyedVectors.load_word2vec_format("data/word2vec_200k.txt")

# Similaridade e Distância Cosseno 

Como comentamos em sala de aula, podemos considerar as palavras como pontos num espaço n-dimensional e podemos examinar a proximidade delas através da similaridade cosseno:

```
$$s = \frac{u \cdot v}{||u|| ||v||}, \textrm{ onde } s \in [-1, 1] $$ 
```

## <font color='blue'>Questão 1 </font>
Palavras [polissemicas](https://pt.wikipedia.org/wiki/Polissemia) e [homônimas](https://pt.wikipedia.org/wiki/Hom%C3%B3nimo) são palavras que possuem mais de um significado. 


Utilizando a função `model.most_similar(positive = palavra1)`, que retorna uma lista das palavras mais similares à palavra1, encontre uma palavra que possua múltiplos significados. Observe que na sua lista de 10 palavras mais similares existam palavras relacionadas a mais de um dos seus significados, lembre-se de consultar sua [documentação](https://radimrehurek.com/gensim/models/keyedvectors.html#gensim.models.keyedvectors.FastTextKeyedVectors.most_similar). 

Por exemplo, a palavra "manga" possui na sua lista de 10 palavras mais similares as palavras "gola" e "lapela" (que estão relacionadas ao significado de manga de uma camiseta) e a palavra "maçã" (que está relacionada ao significado da fruta manga).



In [12]:
# Seu código aqui

model.most_similar("laranja")

[('limão', 0.8328607678413391),
 ('bege', 0.8124987483024597),
 ('azul', 0.8054619431495667),
 ('tangerina', 0.7963930368423462),
 ('gengibre', 0.7915081977844238),
 ('baunilha', 0.7877084612846375),
 ('marrom', 0.7876092195510864),
 ('violeta', 0.784996509552002),
 ('malagueta', 0.7839564085006714),
 ('alface', 0.7821038365364075)]


**<font color='red'> Sua resposta aqui </font>**

Laranja:

 [('limão', 0.8328607678413391), </br>
 ('bege', 0.8124987483024597),</br>
 ('azul', 0.8054619431495667),</br>
 ('tangerina', 0.7963930368423462),</br>
 ('gengibre', 0.7915081977844238),</br>
 ('baunilha', 0.7877084612846375),</br>
 ('marrom', 0.7876092195510864),</br>
 ('violeta', 0.784996509552002),</br>
 ('malagueta', 0.7839564085006714),</br>
 ('alface', 0.7821038365364075)]</br>

# Sinônimos e Antônimos


As vezes é mais intuitivo trabalhar com uma medida de distancia ao invés da similaridade cosseno, para isso vamos utilizar a distancia cosseno que é simplesmente 1 - Similaridade Cosseno.

## <font color='blue'>Questão 2 </font>


Usando a função [`model.distance(palavra1,palavra2)`](https://radimrehurek.com/gensim/models/keyedvectors.html#gensim.models.keyedvectors.FastTextKeyedVectors.distance), encontre 3 palavras onde as palavras p1 e p2 são sinônimas e p1 e p3 são antônimas mas `distance(p1,p3)` < `distance(p1,p2)`.

Proponha uma explicação do porque esse resultado contraintuitivo acontece.






In [24]:
# Seu código aqui

p1 = "lembrar"
p2 = "recordar"
p3 = "esquecer"

par_1 = model.distance(p1,p2)
par_2 = model.distance(p1,p3)

print(f"Distância entre {p1} e {p2}: {par_1}")
print(f"Distância entre {p1} e {p3}: {par_2}")

if par_1 < par_2:
    print(f"distance({p1}, {p2}) < distance({p1}, {p3})")

Distância entre lembrar e recordar: 0.17380636930465698
Distância entre lembrar e esquecer: 0.4047423005104065
distance(lembrar, recordar) < distance(lembrar, esquecer)



**<font color='red'> Sua resposta aqui </font>**

O modelo captura contexto, não lógica semântica. Embeddings pré-treinados capturam relações distributivas, não relações lógicas de oposição. Um modelo não tem uma dimensão interna que diga "este vetor é o oposto daquele". Ele apenas mede a sobreposição de contextos. Como antônimos compartilham contextos de forma massiva, suas distâncias vetoriais ficam pequenas — às vezes menores do que a distância entre sinônimos de registros diferentes.

# Analogias

Existem algumas analogias famosas realizadas por vetores de palavras. O exemplo mais famoso é provavelmente "man : king :: woman : x", onde x é *queen*.

Para formular analogias vamos utilizar a função `most_similar()` que busca as palavras mais similares as listas em  `positive` e mais dissimilares as listadas em  `negative`. Para mais detalhes recomendamos consultar a sua [documentação](https://radimrehurek.com/gensim/models/keyedvectors.html#gensim.models.keyedvectors.FastTextKeyedVectors.most_similar).




In [25]:
model.most_similar(positive=['mulher', 'engenheiro'], negative=['homem'])

[('engenheira', 0.7883446216583252),
 ('investigadora', 0.7415961623191833),
 ('ex-funcionária', 0.7373332977294922),
 ('enfermeira', 0.7346670627593994),
 ('funcionária', 0.7172971367835999),
 ('bibliotecária', 0.7110161781311035),
 ('arquiteta', 0.7099220156669617),
 ('empresária', 0.7055575251579285),
 ('ex-diretora', 0.7055395841598511),
 ('professora', 0.697813868522644)]

## <font color='blue'>Questão 3 </font>
Encontre analogias que funcionam, ou seja, que a palavra esperada está no topo da lista.

Descreva sua analogia na seguinte forma: 
x:y :: a:b



In [61]:
# Seu código aqui
model.most_similar(
    positive=['professor', 'escola'], 
    negative=['doutor']
)

[('faculdade', 0.8312197327613831),
 ('academia', 0.7912424206733704),
 ('cinematografia', 0.7668203711509705),
 ('acadêmia', 0.7625048756599426),
 ('biblioteca', 0.7587317824363708),
 ('pró-associação', 0.7553501129150391),
 ('especialidade', 0.751684308052063),
 ('dramaturgia', 0.7493905425071716),
 ('arquitectura', 0.7422029376029968),
 ('unidade', 0.7420397400856018)]


**<font color='red'> Sua resposta aqui </font>**

Professor está para escola assim como Doutor está para faculdade.

## <font color='blue'>Questão 4 </font>
Encontre analogias que **Não** funcionam.

Descreva o resultado esperado da sua analogia na seguinte forma: 
x:y :: a:b

E indique o valor errado de b encontrado



In [65]:
# Seu código aqui
model.most_similar(
    positive=['brasileiro', 'japão'], 
    negative=['brasil']
)

[('espanhol', 0.8101382851600647),
 ('sulamericano', 0.7973958253860474),
 ('italiano', 0.7795351147651672),
 ('mineiro', 0.7354663610458374),
 ('chileno', 0.720748245716095),
 ('gaúcho', 0.7183815240859985),
 ('moldávio', 0.7140166163444519),
 ('milanista', 0.7120137214660645),
 ('citadino', 0.7107394337654114),
 ('catalão', 0.7005788683891296)]


**<font color='red'> Sua resposta aqui </font>**

A operação brasileiro − brasil + japão caiu numa região genérica de "gentílicos de origem ocidental/latina", não na direção específica "Japão → japonês". 

# Viés e preconceito adquirido

Como estes vetores são aprendidos a partir de documentos produzidos pela nossa sociedade, ele pode vir a capturar alguns preconceitos e desigualdades presentes na nossa sociedade. É importante estar ciente desse viés de nossos vetores e dos seus perigos, aplicações que utilizam esses modelos podem acabar perpetuando e até mesmo exacerbando desigualdades sociais.

Por exemplo, uma analogia problemática capturada:



In [66]:
model.most_similar(positive=['negro', 'rico'], negative=['pobre'])

[('branco', 0.663209080696106),
 ('alegre/rs', 0.6620162725448608),
 ('braga-fc', 0.6464027762413025),
 ('sporting-fc', 0.6254758238792419),
 ('côvo', 0.6254613995552063),
 ('alegre-rs', 0.6199708580970764),
 ('vermelho', 0.612277090549469),
 ('covo', 0.604120671749115),
 ('cirílicos', 0.6022458672523499),
 ('benfica-fc', 0.5965930819511414)]

Note também como diferem as palavras mais semelhantes a homem e mulher:

In [67]:
model.most_similar("homem")

[('monstro', 0.9085395932197571),
 ('bebé', 0.9072304368019104),
 ('indivíduo', 0.9050756096839905),
 ('rapaz', 0.9036115407943726),
 ('mendigo', 0.9007540345191956),
 ('rapazola', 0.8992964029312134),
 ('novelo', 0.8938027620315552),
 ('pássaro', 0.8897998929023743),
 ('cão', 0.8882535099983215),
 ('cãozinho', 0.8869855403900146)]

In [68]:
model.most_similar("mulher")

[('menina', 0.911119282245636),
 ('amiga', 0.9089193344116211),
 ('cadela', 0.9035040140151978),
 ('rapariga', 0.899989902973175),
 ('enfermeira', 0.8974366784095764),
 ('namorada', 0.8954240083694458),
 ('cafetina', 0.8932163119316101),
 ('prostituta', 0.8917951583862305),
 ('garota', 0.8906298279762268),
 ('cadelinha', 0.8902611136436462)]

## <font color='blue'>Questão 5 </font>

Utiliza a função `most_similar()` para encontrar um outro caso de viés adquirido pelos vetores e explique brevemente o tipo de viés encontrado.



In [80]:
# Seu código aqui
model.most_similar("velho")

[('rico', 0.7601529359817505),
 ('azarado', 0.7103042006492615),
 ('antigo', 0.705261766910553),
 ('afamado', 0.6978198885917664),
 ('poderoso', 0.6893930435180664),
 ('querido', 0.6876501441001892),
 ('longevo', 0.6817540526390076),
 ('apetecido', 0.6731494665145874),
 ('opulento', 0.6714193820953369),
 ('afortunado', 0.670701801776886)]


**<font color='red'> Sua resposta aqui </font>**

A palavra "velho" deveria, semanticamente, estar mais próxima de termos como idoso, ancião, decrépito, senil ou experiente. Em vez disso, os vizinhos mais próximos são rico, poderoso, opulento, afamado e querido.
O modelo aprendeu que, no português brasileiro dos textos de treinamento, "velho" não é apenas um marcador de idade — é um marcador de status.

## <font color='blue'>Questão 6 </font>

Qual é a possivel origem desses vieses? Tente explicar como eles podem ter sido capturados pelos vetores de palavras.


**<font color='red'> Sua resposta aqui </font>**

O viés aqui é culturalmente específico do Brasil: o modelo aprendeu que "velho" frequentemente denota riqueza herdada, poder tradicional e afeto paternalista — não apenas deterioração biológica. O corpus refletiu uma sociedade onde a riqueza e o poder se concentram nas mãos de gerações mais velhas, e o embedding espelhou essa realidade social como se fosse uma verdade linguística.